wasmparser: Apply clippy suggestions + add clippy CI job

Robbepop opened the pull request:
This PR adds a `clippy` CI job for `wasmparser` that can be extended to other crates of this workspace in later PRs.
Also the PR applies all `clippy` suggestions to the `wasmparser` crate.

Robbepop commented on crates/wasmparser/src/validator.rs:
```diff
@@ -1792,7 +1791,7 @@ mod arc {
             }
             debug_assert!(Arc::get_mut(&mut self.arc).is_some());
-            Some(unsafe { &mut *(&*self.arc as *const T as *mut T) })
+            Some(self.assert_mut())
```
The downside of this change is that it introduces a runtime check.
The ideal replacement was [`Arc::get_mut_unchecked`](https://doc.rust-lang.org/alloc/sync/struct.Arc.html#method.get_mut_unchecked) but that's still unstable as of now.

alexcrichton replied:
This needs to stay as it was, otherwise this is a mutally recursive function that never returns.

Robbepop replied:
Applied your suggestion and tests are running now again, too. :tada:

Robbepop commented:
I don't understand why the tests are failing. The `wasmparser` tests work fine locally.

Robbepop commented:
@alexcrichton Thanks for the review! Applied your suggestions.

Robbepop commented:
@alexcrichton Would you be interested for a similar PR for the `wast` crate? I'd be down to fix `clippy` lints there as well as adding it to the CI when everything is cleaned up.

alexcrichton commented on crates/wasmparser/src/validator.rs:
```diff
@@ -1792,7 +1791,19 @@ mod arc {
+                // one owner of the `Arc`.
+                // As soon as `Arc::get_mut_unchecked` we should
+                // make use of it instead.
+                #[allow(clippy::cast_ref_to_mut)]
```
I would prefer to have no `#[clippy]` attributes in this crate, can this be removed?

Robbepop replied:
How are we going to satisfy the `clippy` CI then?
Are those `#[clippy]` attributes okay in the crate root? Less ideal though ...

alexcrichton replied:
We don't run clippy in CI in this repository, so there's nothing to satisfy.

Robbepop replied:
Is there a reason for that? With this PR the `wasmparser` crate could profit from having a `clippy` CI job at hand. In fact this PR adds a `clippy` CI job for `wasmparser`.

alexcrichton replied:
Personally I do not believe that clippy adds value to a project. That's my personal opinion, others may feel differently. I don't really want to debate the merits of clippy here at this time, my point is that we don't run clippy and that is not something I would like to change off-hand in a PR.

alexcrichton commented on crates/wasmparser/src/validator.rs:
```diff
@@ -1792,7 +1791,19 @@ mod arc {
+                // `self.owned` is only ever `true` if there is just
+                // one owner of the `Arc`.
+                // As soon as `Arc::get_mut_unchecked` we should
+                // make use of it instead.
```
This `# Safety` comment doesn't really add anything over the structure of the code that's already here, so I don't think it should be added.

Robbepop commented:
I applied your review comments:

- Removed `clippy` CI job for `wasmparser`.
- Removed `Safety` comment.
- Remove attribute to silence a `clippy` warning for unsafe code.
